### PR TITLE
Move to shared config

### LINKED ISSUE #9: Expected empty line between multiple components
I'm getting a **rule-non-nested-empty-line-before** error when I have multiple components in one file

I've got a file like this

```js
import styled from 'styled-component';

const TextA = styled.span`
  color: blue;
`

const TextB = styled.span`
  color: red;
`
```

and got this error
```
7:1  ✖  Expected empty line before non-nested rule   rule-non-nested-empty-line-before
```

I've console.log the extractedCSS and it looks like this
```css
.selector {
  color: blue;
}  /* <----- missing a newline here */
.selector {
  color: red;
}
```

I'm using:
* stylelint-config-standard@15.0.0
* stylelint-processor-styled-components@0.0.4
* stylelint@7.6.0

### LINKED ISSUE #45: Add external stylelint config
Which can be extended by users of this processor. It'll mean that we won't have to modify the stylelint result anymore, and that people can still override any of the rules we define (should they want to do so).

It'll need to be a separate library, but still needs to be linked to a particular release of this lib, so specifying the config as a peerDependency would probably be best.

Ref #9

### PR DESCRIPTION
**Pull request by ismay**
Closes #45
Closes #9 

This'll be a breaking change. It can break builds because we're now prohibiting the use of vendor-prefixed properties and values, and a failure to include the config as a dependency could potentially break builds as well (unfixable rules won't be ignored anymore).

- [x] Add as peerDependency in package.json
- [x] Remove ignored rules, don't modify stylelint result (too much :)
- [x] Update readme

### CODE REVIEW

**Comment by mxstbr** [comment]
Wait, what's missing here? Is this done? We should release this with 1.0!

**Comment by ismay** [comment]
@mxstbr Definitely. I was waiting for the more significant refactors, but yeah this should definitely be a part of v1 as it'll be a breaking change. I'll start a pr to allow for feedback. But it should be quite simple.

**Comment by ismay** [comment]
Rebased it on master. The package-lock update is actually unrelated to this pr, but something that always happens for me with greenkeeper lockfile updates. For some reason a regular `npm install` changes the lockfile, even though no new dependencies were installed. Happens only for me when greenkeeper's updated the lockfile, weird.

Anyway. I've added stylelint-config-styled-components-processor as a peerDependency. One thing I've been thinking about is where the config should be tested. Way I see it, is that we could just add some simple tests to the stylelint-config-styled-components-processor repo, just to verify that it's ignoring and enforcing the correct rules.

Other than that the todo's above are all I think should be changed. Let me know if there's anything that I've overlooked!

**Comment by emilgoldsmith** [comment]
> For some reason a regular npm install changes the lockfile, even though no new dependencies were installed. Happens only for me when greenkeeper's updated the lockfile, weird.

It could be because the `greenkeeper-lockfile` repo uses a different version of npm than you / the CI uses a different version than you, not something I've double checked but I know some of the newer versions of NPM do these optimizations to the lockfile, so could be that.
But yeah weird for sure.

**Comment by ismay** [comment]
Updated typescript to the version that typescript-eslint-parser currently supports (2.4.0). I could take it out since it isn't specifically related to this pr, but I was getting a bunch of warnings during testing. Still feels weird to bundle typescript as a dependency btw, when it seems more appropriate as a peerDependency. Maybe good to discuss that in another issue, don't know how you feel about that?

But back to the matters at hand: index.js has been updated to no longer ignore any stylelint rules. @emilgoldsmith notice anything I might have missed? Basically I think this is it, and we just need to update the readme.

**Comment by ismay** [comment]
Great thing is that now we're also not relying too much on the internal structure of the stylelintResult anymore. Only thing we're doing is replace brace with backtick, which is quite safe.

**Comment by emilgoldsmith** [comment]
Yeah this is a really cool idea @ismay! I'll just take a bit of time to look at this and the repo you created to understand exactly what it is we're doing so I can give better feedback but I'll get back to you ASAP!

Also, I was just thinking, since we probably don't want to for example release this one before 1.0 (unless we just publish a 0.4.0 of course) we could make a 1.0 branch now that we merge this and other breaking stuff like my PR into until we're ready to publish and then merge it into master?

**Comment by ismay** [comment]
> Yeah this is a really cool idea @ismay! I'll just take a bit of time to look at this and the repo you created to understand exactly what it is we're doing so I can give better feedback but I'll get back to you ASAP!

Cool thanks!

> Also, I was just thinking, since we probably don't want to for example release this one before 1.0 (unless we just publish a 0.4.0 of course) we could make a 1.0 branch now that we merge this and other breaking stuff like my PR into until we're ready to publish and then merge it into master?

Yeah sounds like a good idea 👍

**Comment by emilgoldsmith** [comment]
> Updated typescript to the version that typescript-eslint-parser currently supports (2.4.0). I could take it out since it isn't specifically related to this pr, but I was getting a bunch of warnings during testing.

Yeah I was getting that as well, it was so annoying, thanks for updating it! I don't think it should be a problem doing it in this PR though it could of course be seen as cleaner doing it in a different one.

> Still feels weird to bundle typescript as a dependency btw, when it seems more appropriate as a peerDependency. Maybe good to discuss that in another issue, don't know how you feel about that?

Yeah why is it it's a dependency by the way? It shouldn't need to be a peerDependency either should it? As in, if you don't use Typescript you shouldn't need it, couldn't we just make it a devDependency since the only reason we have it here is for testing, right? Or am I missing something?

**Comment by ismay** [comment]
> Yeah why is it it's a dependency by the way? It shouldn't need to be a peerDependency either should it? As in, if you don't use Typescript you shouldn't need it, couldn't we just make it a devDependency since the only reason we have it here is for testing, right? Or am I missing something?

I think we did that because typescript-eslint-parser throws if typescript can't be found (or complained anyway). Which is annoying for people who don't use our processor for typescript as they have no need for typescript in their project. So we're basically stuck with this situation unless it can somehow be fixed on typescript-eslint-parser's end (by not doing the check until it's actually instantiated or something, don't know what they're doing exactly).

**Comment by emilgoldsmith** [comment]
> I think we did that because typescript-eslint-parser throws if typescript can't be found (or complained anyway). 

Hmm that's a pain <.<. When I'm done reviewing this which should be in a sec I'll just look at our code regarding Typescript as I have never looked at that, mostly out of curiosity, would be awesome if we could remove that dependency somehow for sure though it doesn't sound like it.

File: src/index.js [hunk 1]
```diff
@@ -1,16 +1,6 @@
 const path = require('path')
 const parse = require('./parsers/index')
 // TODO Fix ampersand in selectors
```
Comment (emilgoldsmith) [inline src/index.js]: What is it we mean here? Is this still a TODO?

File: src/index.js [hunk 2]
```diff
@@ -30,23 +20,15 @@ module.exports = (/* options */) => ({
   // Fix sourcemaps
   result(stylelintResult, filepath) {
     const lineCorrection = sourceMapsCorrections[filepath]
-    const newWarnings = stylelintResult.warnings.reduce((prevWarnings, warning) => {
-      if (ignoredRules.includes(warning.rule)) return prevWarnings
-      const correctedWarning = Object.assign(warning, {
+    const warnings = stylelintResult.warnings.map(warning =>
+      Object.assign(warning, {
         // Replace "brace" with "backtick" in warnings, e.g.
         // "Unexpected empty line before closing backtick" (instead of "brace")
```
Comment (emilgoldsmith) [inline src/index.js]: This wouldn't be correct if it you had a declaration block inside of your backticks, right? Maybe we should do some check and only do that replace if it's on line 0 or the last line of the block?

File: src/index.js [hunk 3]
```diff
@@ -30,23 +20,15 @@ module.exports = (/* options */) => ({
   // Fix sourcemaps
   result(stylelintResult, filepath) {
     const lineCorrection = sourceMapsCorrections[filepath]
-    const newWarnings = stylelintResult.warnings.reduce((prevWarnings, warning) => {
-      if (ignoredRules.includes(warning.rule)) return prevWarnings
-      const correctedWarning = Object.assign(warning, {
+    const warnings = stylelintResult.warnings.map(warning =>
+      Object.assign(warning, {
```
Comment (emilgoldsmith) [inline src/index.js]: Hmm for good form I might not mutate `warning` and do `Object.assign({}, warning, {`. Also see my comment on line 32 below.

File: src/index.js [hunk 4]
```diff
@@ -30,23 +20,15 @@ module.exports = (/* options */) => ({
...
-    }, [])
-
-    if (newWarnings.length === 0) {
-      // eslint-disable-next-line no-param-reassign
-      stylelintResult.errored = false
-    }
+    )
 
-    return Object.assign(stylelintResult, { warnings: newWarnings })
+    return Object.assign(stylelintResult, { warnings })
```
Comment (emilgoldsmith) [inline src/index.js]: As we currently mutate `warning` in the `.map` above this is actually unnecessary right? I think with current mutations this is actually the same as just returning `stylelintResult`. I understand that's not as clear though, so maybe I'd recommend just not mutating and following my suggestion above and then also not mutating here? As in again add a first argument `{}`.

File: src/index.js (hunk 1 above)
Comment (ismay) [inline src/index.js]: Yes I was also curious about that. @mxstbr Can we remove this? Was this related to the stylelint rules we wanted to enforce?

**Comment by emilgoldsmith** [comment]
I just created a 1.0 branch and change the base, I'll go change the base of my PR now as well :)

File: src/index.js (hunk 1 above)
Comment (mxstbr) [inline src/index.js]: No idea anymore haha, let's get rid of it!

File: src/index.js (hunk 2 above)
Comment (mxstbr) [inline src/index.js]: Yeah there's definitely a lot more stuff we could/should do with the warnings in the future! Could get a bit smarter :blush:

File: src/index.js (hunk 4 above)
Comment (ismay) [inline src/index.js]: You're right, fixed it as per your suggestions

File: src/index.js (hunk 2 above)
Comment (emilgoldsmith) [inline src/index.js]: Okay I'll open a quick issue to keep it on our radar, but it's definitely not high prio :)

**Comment by ismay** [comment]
I've created https://github.com/styled-components/stylelint-processor-styled-components/issues/99 btw, as a reminder of the typescript dependency discussion.

**Comment by ismay** [comment]
> Looks great, so now we just need to update the README, right?

Yep, and I'll doublecheck whether the shareable config has been published on npm.

**Comment by emilgoldsmith** [comment]
@ismay I just reverted the Typescript version here as I thought now we have the #101 PR we can just as well handle it there, yeah?

**Comment by ismay** [comment]
> @ismay I just reverted the Typescript version here as I thought now we have the #101 PR we can just as well handle it there, yeah?

Yeah sounds good 👍

**Comment by emilgoldsmith** [comment]
Looks good to me! I'm a big fan of this PR, let's merge it in :D

**Pull request merged**